### PR TITLE
Do not upgrade all Elasticsearch nodes of a given tier at once

### PR DESCRIPTION
**Pull request by pebrc**
Fixes https://github.com/elastic/cloud-on-k8s/issues/1753

Introduces a new upgrade predicate to prevent deletion of all the nodes of a given tier. Tier to be understood here as a generalisation of the existing "data tier" notion to all non-master roles. 

Master nodes (and voting only nodes) are excluded as we are handling those already in existing predicates and generally apply much stricter rules to them (just one master node at a time can be upgraded no matter how permissive the change budget)

The focus here is on the existing nodes in the cluster and not on the expected nodes. My rationale is that the purpose of this predicate is to prevent degradation of the existing service for which the future state of the service is irrelevant. 

I have tested a few scenarios manually: 

* single node tiers which are treated as special case
* node role changes (addition/removal of node roles to existing nodeSets)
* misconfiguration scenarios (configuration change leads to pending or crash looping Pods

I think the PR could potentially benefit from capturing these scenarios in additional automated tests but I thought I raise it already to get some initial feedback.

### CODE REVIEW

**Comment by pebrc** [comment]
full pr build

**Comment by pebrc** [comment]
run full pr build

**Comment by pebrc** [comment]
> pods is forbidden: User "system:serviceaccount:e2e-2uhwl:e2e-2uhwl" cannot list resource "pods" in API group "" in the namespace "e2e-2uhwl-mercury"

Odd failure 😕  e2e RBAC should include those permissions.

**Comment by pebrc** [comment]
run full pr build

**Review by barkbay — APPROVED** [review]
LGTM

File: pkg/controller/elasticsearch/driver/upgrade_predicates.go [hunk 1]
```diff
@@ -556,6 +558,63 @@ var predicates = [...]Predicate{
...
+					}
+				}
+			}
+
+			for _, role := range nonMasterRoles {
+				if role.HasValue(true, candidate.Labels) {
+					healthy := healthyPodRoleHisto[role]
+					current := currentPodRoleHisto[role]
+					if current == 1 {
+						// only Pod with this role or all unhealthy: OK to delete
```
[the remark this answers is not in the document]
Comment (pebrc) [reply pkg/controller/elasticsearch/driver/upgrade_predicates.go:600]: Yes, I think that is a good idea.

**Pull request merged**